lxd/container_lxc.go

```diff
@@ -3722,30 +3735,47 @@ func (c *containerLXC) createUnixDevice(m shared.Device) (string, error) {
 
 	// Clean any existing entry
 	if shared.PathExists(devPath) {
+		if runningInUserns {
+			syscall.Unmount(devPath, syscall.MNT_DETACH)
+		}
+
 		err = os.Remove(devPath)
 		if err != nil {
 			return "", fmt.Errorf("Failed to remove existing entry: %s", err)
 		}
 	}
 
 	// Create the new entry
-	if err := syscall.Mknod(devPath, uint32(mode), minor|(major<<8)); err != nil {
-		return "", fmt.Errorf("Failed to create device %s for %s: %s", devPath, m["path"], err)
-	}
+	if !runningInUserns {
+		if err := syscall.Mknod(devPath, uint32(mode), minor|(major<<8)); err != nil {
+			return "", fmt.Errorf("Failed to create device %s for %s: %s", devPath, m["path"], err)
+		}
 
-	if err := os.Chown(devPath, uid, gid); err != nil {
-		return "", fmt.Errorf("Failed to chown device %s: %s", devPath, err)
-	}
+		if err := os.Chown(devPath, uid, gid); err != nil {
+			return "", fmt.Errorf("Failed to chown device %s: %s", devPath, err)
+		}
 
-	// Needed as mknod respects the umask
-	if err := os.Chmod(devPath, mode); err != nil {
-		return "", fmt.Errorf("Failed to chmod device %s: %s", devPath, err)
-	}
+		// Needed as mknod respects the umask
+		if err := os.Chmod(devPath, mode); err != nil {
+			return "", fmt.Errorf("Failed to chmod device %s: %s", devPath, err)
+		}
 
-	if c.idmapset != nil {
-		if err := c.idmapset.ShiftFile(devPath); err != nil {
-			// uidshift failing is weird, but not a big problem.  Log and proceed
-			shared.Debugf("Failed to uidshift device %s: %s\n", m["path"], err)
+		if c.idmapset != nil {
+			if err := c.idmapset.ShiftFile(devPath); err != nil {
+				// uidshift failing is weird, but not a big problem.  Log and proceed
+				shared.Debugf("Failed to uidshift device %s: %s\n", m["path"], err)
+			}
+		}
+	} else {
+		f, err := os.Create(devPath)
```

Comment:
So specifically I think the problem is that since our queue is size limited (to 10 entries), further writes will be blocked until existing entries have been processed.

But since the go routine processing the queue also needs to grab those locks, we were getting into a complete deadlock whenever you'd hit the case where the queue was full yet you still had more entries to append to it.

My change has us release the lock immediately after we're done with it, and most importantly, we no longer hold the lock when appending to the queue, so if we get stuck at that spot, the go routines can still do their work.